File: motio_client/lib/models/debt.dart
import 'transaction.dart';
import 'user.dart';

class Debt {
  final int id;
  final User user1;
  final User user2;
  final List<Transaction> transactionHistories;
  final double balance;

  Debt({
    required this.id,
    required this.user1,
    required this.user2,
    required this.transactionHistories,
    required this.balance,
  });

  factory Debt.fromJson(Map<String, dynamic> json) {
    return Debt(
      id: json['id'],
      user1: User.fromJson(json['user1']),
      user2: User.fromJson(json['user2']),
      transactionHistories: (json['transactionHistories'] as List)
          .map((item) => Transaction.fromJson(item, null))
          .toList(),
      balance: json['balance'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user1': user1.toJson(),
      'user2': user2.toJson(),
      'transactionHistories':
      transactionHistories.map((item) => item.toJsonWithoutDebt()).toList(),
      'balance': balance,
    };
  }
}

File: motio-commons/src/main/java/com/motio/commons/model/Debt.java
package com.motio.commons.model;

import com.fasterxml.jackson.annotation.JsonManagedReference;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.util.LinkedList;
import java.util.List;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(name = "debts")
public class Debt {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne
    @JoinColumn(name = "user_1_id", nullable = false)
    private User user1;

    @ManyToOne
    @JoinColumn(name = "user_2_id", nullable = false)
    private User user2;

    @OneToMany(mappedBy = "debt", cascade = CascadeType.ALL, orphanRemoval = true)
    @JsonManagedReference
    private List<Transaction> transactionHistories = new LinkedList<>();

    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal balance;
}
